### PR TITLE
Added support for using HTML inside validation error messages

### PR DESCRIPTION
**Pull request by mpscholten**
Fixes https://github.com/digitallyinduced/ihp/issues/1093

### CODE REVIEW

File: IHP/ValidationSupport/Types.hs [hunk 1]
```diff
@@ -2,33 +2,110 @@ module IHP.ValidationSupport.Types where
...
+--
+-- __Example:__
+--
+-- >>> record |> attachFailure #email "should be a valid email"
+-- User { .., meta = MetaBag { .., annotations = [ ("email", TextViolation "should be a valid email") ] } }
+--
+-- You can use this together with 'getValidationFailure'
+--
+-- > user
+-- >     |> attachFailure #email"cannot be empty"
```
Comment (amitaibu) [inline IHP/ValidationSupport/Types.hs:48]:
```suggestion
-- >     |> attachFailure #email "cannot be empty"
```

File: IHP/ValidationSupport/Types.hs [hunk 2]
```diff
@@ -2,33 +2,110 @@ module IHP.ValidationSupport.Types where
...
+--
+-- __Example:__
+--
+-- >>> record |> attachFailureHtml #email [hsx|should be a valid email. <a href="https://example.com/docs#email">Check out the documentation</a>|]
+-- User { .., meta = MetaBag { .., annotations = [ ("email", HtmlViolation "should be a valid email. <a href="https://example.com/docs#email">Check out the documentation</a>") ] } }
+--
+-- You can use this together with 'getValidationViolation'
+--
+-- > user
+-- >     |> attachFailure #email"cannot be empty"
```
Comment (amitaibu) [inline IHP/ValidationSupport/Types.hs:68]:
```suggestion
-- >     |> attachFailure #email "cannot be empty"
```

File: IHP/ValidationSupport/Types.hs [hunk 3]
```diff
@@ -2,33 +2,110 @@ module IHP.ValidationSupport.Types where
...
+    where
+        renderedHtml = message
+                |> Blaze.renderHtml
+                |> cs
+{-# INLINE attachFailureHtml #-}
+
+-- | Returns the validation failure for a field or Nothing
+--
+-- > user
+-- >     |> attachFailure #email"cannot be empty"
```
Comment (amitaibu) [inline IHP/ValidationSupport/Types.hs:83]:
```suggestion
-- >     |> attachFailure #email "cannot be empty"
```

File: Guide/validation.markdown [hunk 4]
```diff
@@ -323,6 +323,24 @@ post
...
 This record now has a validation error attached to its title.
 
+#### Attaching Errors with HTML
+
+If you try to use HTML code within `attachFailure`, the HTML code will be escaped and not rendered as expected:
+
+```haskell
+post
+    |> attachFailure #title "Invalid value. <a href="https://example.com/docs">Check the documentation</a>"
+    -- Link will not be clickable as it's going through escaping
```
Comment (amitaibu) [inline Guide/validation.markdown:333]:
```suggestion
    -- Link will not be clickable as the HTML is escaped
```

**Review by amitaibu — COMMENTED** [review]
Thanks. I'm not knowledgeable enough about IHP, but a couple of missing space nitpicks :smile_cat:

**Comment by mpscholten** [comment]
Thanks :) Good points, all applied 👍

File: IHP/ModelSupport.hs [hunk 5]
```diff
@@ -600,9 +600,15 @@ ids :: (HasField "id" record id) => [record] -> [id]
...
+-- | The error message of a validator can be either a plain text value or a HTML formatted value
+data Violation
+    = TextViolation { message :: !Text } -- ^ Plain text validation error, like "cannot be empty"
+    | HtmlViolation { message :: !Text } -- ^ HTML formatted, already pre-escaped validation error, like "Invalid, please <a href="http://example.com">check the documentation</a>"
+    deriving (Eq, Show)
+
 -- | Every IHP database record has a magic @meta@ field which keeps a @MetaBag@ inside. This data structure is used e.g. to keep track of the validation errors that happend.
 data MetaBag = MetaBag
-    { annotations            :: ![(Text, Text)] -- ^ Stores validation failures, as a list of (field name, error) pairs. E.g. @annotations = [ ("name", "cannot be empty") ]@
+    { annotations            :: ![(Text, Violation)] -- ^ Stores validation failures, as a list of (field name, error) pairs. E.g. @annotations = [ ("name", "cannot be empty") ]@
```
Comment (amitaibu) [inline IHP/ModelSupport.hs:611]: Not sure, but maybe this should be:

```suggestion
    { annotations            :: ![(Text, Violation)] -- ^ Stores validation failures, as a list of (field name, error) pairs. E.g. @annotations = [ ("name", TextViolation "cannot be empty") ]@
```

Comment (mpscholten) [reply IHP/ModelSupport.hs:611]: Good catch, fixed via https://github.com/digitallyinduced/ihp/commit/70738b2541a4331c55ed74f7bfdb979ae990f815 👍

**Pull request merged**